e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js

```diff
@@ -0,0 +1,524 @@
+const VIEWPORTS = [
+  { width: 375, height: 667, openSidebar: false },
+  { width: 820, height: 800, openSidebar: true },
+  { width: 820, height: 800, openSidebar: false },
+  { width: 1200, height: 800, openSidebar: true },
+  { width: 1440, height: 800, openSidebar: true },
+  { width: 1440, height: 800, openSidebar: false },
+];
+
+const SCALAR_QUESTION = {
+  name: "31628 Question - This is a rather lengthy question name",
+  description: "This is a rather lengthy question description",
+  query: {
+    "source-table": ORDERS_ID,
+    aggregation: [["count"]],
+  },
+  display: "scalar",
+};
+
+const SCALAR_QUESTION_CARDS = [
+  { cards: createCardsRow({ size_y: 2 }), name: "cards 2 cells high" },
+  { cards: createCardsRow({ size_y: 3 }), name: "cards 3 cells high" },
+  { cards: createCardsRow({ size_y: 4 }), name: "cards 4 cells high" },
+  CARDS_SIZE_1X,
+];
+
+const SMART_SCALAR_QUESTION = {
+  name: "31628 Question - This is a rather lengthy question name",
+  description: "This is a rather lengthy question description",
+  query: {
+    "source-table": ORDERS_ID,
+    aggregation: [["count"]],
+    breakout: [
+      [
+        "field",
+        ORDERS.CREATED_AT,
+        {
+          "base-type": "type/DateTime",
+          "temporal-unit": "month",
+        },
+      ],
+    ],
+  },
+  display: "smartscalar",
+};
+
+const SMART_SCALAR_QUESTION_CARDS = [
+  { cards: createCardsRow({ size_y: 2 }), name: "cards 2 cells high" },
+  { cards: createCardsRow({ size_y: 3 }), name: "cards 3 cells high" },
+  { cards: createCardsRow({ size_y: 4 }), name: "cards 4 cells high" },
+];
+
+describe("issue 31628", () => {
+  describe("display: scalar", () => {
+    const descendantsSelector = [
+      "[data-testid='scalar-container']",
+      "[data-testid='scalar-title']",
+      "[data-testid='scalar-description']",
+    ].join(",");
+
+    VIEWPORTS.forEach(({ width, height, openSidebar }) => {
+      SCALAR_QUESTION_CARDS.forEach(({ cards, name }) => {
+        const sidebar = openSidebar ? "sidebar open" : "sidebar closed";
+
+        describe(`${width}x${height} - ${sidebar} - ${name}`, () => {
+          beforeEach(() => {
+            restore();
+            cy.viewport(width, height);
+            cy.signInAsAdmin();
+            setupDashboardWithQuestionInCards(SCALAR_QUESTION, cards);
+
+            if (openSidebar) {
+              openNavigationSidebar();
+            }
+          });
+
+          it(`should render descendants of a 'scalar' without overflowing it (metabase#31628)`, () => {
+            assertDescendantsNotOverflowDashcards(descendantsSelector);
+          });
+        });
+      });
+    });
+
+    describe("1x2 card", () => {
+      beforeEach(() => {
+        restore();
+        cy.signInAsAdmin();
+        setupDashboardWithQuestionInCards(SCALAR_QUESTION, [
+          { size_x: 1, size_y: 2, row: 0, col: 0 },
+        ]);
+      });
+
+      it("should truncate value and show value tooltip on hover", () => {
```

Comment:
Separating tests could make it easier to understand however with the application our size, this increases the test time significantly.
#### This PR which has 3 tests in `1x2 card`
![Screenshot 2023-06-29 at 1 20 52 PM](https://github.com/metabase/metabase/assets/1937582/71bd0acf-dc17-4b62-9e2a-7da01831fb00)

#### If we assert everything in the same test
![Screenshot 2023-06-29 at 1 21 55 PM](https://github.com/metabase/metabase/assets/1937582/614c1eee-0ad6-4aa7-81da-113c748c907c)

Especially when each test doesn't make any changes to the application it's safe to assert different things in the same test to put less burden on our CI runners.

Reply:
> Separating tests could make it easier to understand however with the application our size

"this becomes even more important" is another good ending to that sentence. Yeah, it's a balance.

But it feels wrong to sacrifice quality of tests because of the infrastructure. I'd rather improve the infrastructure.

----

@WiNloSt I didn't get from your comment whether you're suggesting to do this or just making me aware.
Please let me know if you think I should change it.

Reply:
I was suggesting you combine all the assertions into a single test, sorry that it wasn't clear. Looping @nemanjaglumac in, since he has the most insights to our CI and he suggested me the same thing.

> But it feels wrong to sacrifice quality of tests because of the infrastructure. I'd rather improve the infrastructure.

you made a good point though.

Reply:
I started doing it but it starts to look very cloudy and I can't follow it anymore:
- Even though I wrote this code, it's hard for me to say where a test begins and ends, especially if there were more than 3 `it` calls
- Variables clash - after changing viewport size we often need to reacquire reference to the DOM element and we'd have to go with `let`
- Tests are supposed to be isolated and independent - the suggestion goes against it

IMO we should not be doing this.

Reply:
> We can revisit the test later once Nemanja has provided some more insights for us.

The bad news is that e2e tests are timing out, so I might do it after all: https://github.com/metabase/metabase/actions/runs/5409602915/jobs/9830378200?pr=31722

Reply:
🙃

Reply:
Optimized and added a comment: faadfa0c10132767c887922441fe0cd0b5cb1413
I used `/** */` comments to make test cases more prominent.

Reply:
Now we pray to the CI god 🤞

Reply:
In the end I had to adjust the tests because of the optimization:
- after the change, a single test was testing 2 tooltips, so I needed to trigger `mouseleave` to make sure at most 1 tooltip is displayed at a time
- test cases with multiple `cy.viewport()` calls were flaky unless they were wrapped in `it` block, so I wrapped them back

This change made tests very difficult to debug.
I hope we will improve our infrastructure so that we don't need to compromise on quality.